File: examples/src/bin/foci_stm.rs
// Circular foci STM at 1 Hz. Run with: cargo xtask example foci_stm

use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry, Vector3, offset};
use autd3_rs::units::{Hz, mm};
use autd3_rs::value::Intensity;
use autd3_rs::{Client, ClientConfig, FociStm, FociStmOption, Silencer, circle};
use autd3_rs_link_ethercrab::EtherCrabLinkOption;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("info")),
        )
        .init();

    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(
        &geometry,
        EtherCrabLinkOption::default(),
        ClientConfig::default(),
    )
    .await?;

    println!("devices: {}", client.num_devices());

    // 200-point circle of radius 30 mm, 150 mm above the array center.
    let center = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let points = circle(center, 30.0 * mm, 200, Vector3::z_axis(), Intensity::MAX);

    let mut builder = client.datagram_builder();
    builder.push(Silencer::default()).push(FociStm::new(
        1.0 * Hz,
        &points,
        FociStmOption::default(),
    ));
    let datagrams = builder.build()?;
    let mut pending = Vec::with_capacity(datagrams.len());
    for frame in &datagrams {
        pending.push(client.send(frame).await?);
    }
    for response in pending {
        response.await?.check()?;
    }

    println!("running a 1 Hz circular foci STM — press Ctrl+C to stop");
    tokio::signal::ctrl_c().await?;

    client.stop().await?;
    client.close().await?;
    Ok(())
}

File: bindings/python/examples/foci_stm.py
"""
Sweeps a focus around a 30 mm circle at 1 Hz using FociStm.

Run with: cargo xtask py example foci_stm
"""

import asyncio
import signal

import numpy as np

import autd3
import autd3_link_echocat as echocat
from autd3.units import Hz


async def main() -> None:
    geometry = autd3.geometry.Geometry([autd3.geometry.Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client = await autd3.Client.open(
        geometry,
        echocat.EchocatLinkOption(),
        autd3.ClientConfig(),
    )

    print("devices:", client.num_devices())

    center = geometry.center() + np.array([0.0, 0.0, 150.0])
    samples = []
    autd3.commands.circle(center, 30.0, 200, [0.0, 0.0, 1.0], autd3.value.Intensity.MAX, samples)
    stm = autd3.commands.FociStm(1.0 * Hz, samples, autd3.commands.FociStmOption())

    builder = client.datagram_builder()
    builder.push(autd3.commands.SetSilencer())
    builder.push(stm)
    for frame in builder.build():
        await client.send_checked(frame)

    print("sweeping a focus around a 30 mm circle at 1 Hz — press Ctrl+C to stop")
    stop = asyncio.Event()
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, stop.set)
    await stop.wait()

    await client.stop()
    await client.close()


if __name__ == "__main__":
    asyncio.run(main())
